ruby/ql/lib/codeql/ruby/frameworks/ActionDispatch.qll

```diff
@@ -0,0 +1,907 @@
+    (
+      not exists(m.getKeywordArgument("only"))
+      or
+      exists(Expr only | only = m.getKeywordArgument("only") |
+        [only.(ArrayLiteral).getElement(_), only.(StringlikeLiteral)].getValueText() = action
+      )
+    ) and
+    // Respect the `except` keyword argument, which removes actions from the default set.
+    (
+      not exists(m.getKeywordArgument("except"))
+      or
+      exists(Expr except | except = m.getKeywordArgument("except") |
+        [except.(ArrayLiteral).getElement(_), except.(StringlikeLiteral)].getValueText() != action
+      )
+    )
+  }
+
+  /**
+   * Holds if the (resource, method, path, action) combination would be generated by a call to `resources :<resource>`.
+   */
+  bindingset[resource]
+  private predicate isDefaultResourceRoute(
+    string resource, string method, string path, string action
+  ) {
+    action = "create" and
+    (method = "post" and path = "/" + resource)
+    or
+    action = "index" and
+    (method = "get" and path = "/" + resource)
+    or
+    action = "new" and
+    (method = "get" and path = "/" + resource + "/new")
+    or
+    action = "edit" and
+    (method = "get" and path = "/" + resource + ":id/edit")
+    or
+    action = "show" and
+    (method = "get" and path = "/" + resource + "/:id")
+    or
+    action = "update" and
+    (method in ["put", "patch"] and path = "/" + resource + "/:id")
+    or
+    action = "destroy" and
+    (method = "delete" and path = "/" + resource + "/:id")
+  }
+
+  /**
+   * Holds if the (resource, method, path, action) combination would be generated by a call to `resource :<resource>`.
+   */
+  bindingset[resource]
+  private predicate isDefaultSingularResourceRoute(
+    string resource, string method, string path, string action
+  ) {
+    action = "create" and
+    (method = "post" and path = "/" + resource)
+    or
+    action = "new" and
+    (method = "get" and path = "/" + resource + "/new")
+    or
+    action = "edit" and
+    (method = "get" and path = "/" + resource + "/edit")
+    or
+    action = "show" and
+    (method = "get" and path = "/" + resource)
+    or
+    action = "update" and
+    (method in ["put", "patch"] and path = "/" + resource)
+    or
+    action = "destroy" and
+    (method = "delete" and path = "/" + resource)
+  }
+
+  /**
+   * Extract the controller from a Rails routing string
+   * ```
+   * extractController("posts#show") = "posts"
+   * ```
+   */
+  bindingset[input]
+  private string extractController(string input) { result = input.regexpCapture("([^#]+)#.+", 1) }
+
+  /**
+   * Extract the action from a Rails routing string
+   * ```
+   * extractController("posts#show") = "show"
+   */
+  bindingset[input]
+  private string extractAction(string input) { result = input.regexpCapture("[^#]+#(.+)", 1) }
+
+  /**
+   * A basic pluralizer for English strings.
+   * photo => photos
+   * story => stories
+   * TODO: remove?
```

Comment:
This is unused, which is causing the qltest job failure. I guess you can just remove it?

Reply:
Yeah I guess so. I was sure it would be necessary at some point, but apparently not!